app/services/lms_sync_service.py

```diff
@@ -91,6 +104,197 @@ async def sync_assignments(self):
                 )
         
         return canvas_assignments
+    
+    async def _pair_db_canvas_users(self, db_users: list[UserModel], canvas_users: list[dict], canvas_user_pids: list[str]):
+        # User exists in Canvas but not DB = create
+        # User exists in DB but not Canvas = delete
+        # User exists in both DB and Canvas = update
+        pairs = []
+        
+        async def process_canvas_user(canvas_user) -> DatabaseLMSResourcePair | None:
+            pid, email, name = canvas_user.get("sis_user_id"), canvas_user.get("email"), canvas_user.get("name")
+            if pid is None or email is None or name is None:
+                print("Skipping over pending user", name or email or "<unknown>")
+                return None
+
+            try:
+                print("getting user info for ", pid)
+                user_info = self.ldap_service.get_user_info(pid)
+                print(pid, "->", user_info.onyen)
+            except UserNotFoundException:
+                print("Skipping over student not in LDAP: ", pid or email or "<unknown>")
+                return None
+
+            try:
+                user = await self.user_service.get_user_by_onyen(user_info.onyen)
+                return DatabaseLMSResourcePair(db_resource=user, lms_resource=canvas_user, user_info=user_info)
+
+            except UserNotFoundException:
+                return DatabaseLMSResourcePair(db_resource=None, lms_resource=canvas_user, user_info=user_info)
+        
+
+        # We begin by checking if there are any users that have been deleted from Canvas.
+        # This is 1 DB call per user so doesn't really need optimization.
+        for db_user in db_users:
+            pid = await self.canvas_service.get_pid_from_onyen(db_user.onyen)
+            if pid not in canvas_user_pids:
+                pairs.append(DatabaseLMSResourcePair(db_resource=db_user, lms_resource=None))
+        
+        # Then we process Canvas users
+        pairs += await asyncio.gather(*[process_canvas_user(u) for u in canvas_users])
+
+        return [p for p in pairs if p is not None]
+    
+    async def _pair_db_canvas_assignments(self, db_assignments: list[AssignmentModel], canvas_assignments: list[dict]):
```

Comment:
I feel like this `dict` should probably be a Model instead of just a dict

Reply:
You can type the canvas api interfaces. That would take a long time.

Reply:
Why can't you just create a type that has the fields you want to pull in from the response, and drop the fields that you don't care about?

Reply:
I can do that, if you want

Reply:
Sure